File: lib/to_mdx/parsers/open_document/office/automatic_styles.rb
# frozen_string_literal: true

require_relative "../base"

module ToMdx
  module Parsers
    module Office
      class AutomaticStyles < Base
        def initialize(element = nil, context = {})
          super(element, context)

          # TODO if colliding keys, merge the hash values instead of clobbering the key
          context[:styles]&.merge!(to_css)
        end

        def empty?
          false
        end

        def to_mdx
          context[:styles].to_mdx
        end

        def to_css
          children.map(&:to_css).reduce({}, :merge)
        end

        register_for "office:automatic-styles"
      end
    end
  end
end

File: src/parsers/open-document/office/automatic-styles.ts
import type { StyleProperties } from "../../../styles.ts";
import { BaseElement, type ParseContext } from "../../base_element.ts";

export class AutomaticStyles extends BaseElement {
  constructor(element: Element, context: ParseContext, parent: BaseElement | null) {
    super(element, context, parent);
    context.styles.merge(this.toCss());
  }

  empty(): boolean {
    return false;
  }

  toMdx(): string {
    return this.context("styles").toMdx();
  }

  toString(): string {
    return "";
  }

  toCss(): Record<string, StyleProperties> {
    return this.children.reduce(
      (acc, child) => ({
        ...acc,
        // TODO avoid having to this
        ...(child as unknown as { toCss: () => Record<string, string | undefined> }).toCss(),
      }),
      {},
    );
  }

  static {
    BaseElement.registerFor.call(this, "office:automatic-styles");
  }
}
